### ISSUE TITLE
`--headless` doesn't have any effect

### ISSUE THREAD

**Issue by mihe** [godotengine/godot#94755]
### Tested versions

Reproducible in: 4.3.rc [e343dbbcc1030f04dc5833f1c19d267a17332ca9]

### System information

Windows 11 (10.0.22631)

### Issue description

Adding `--headless` to the command-line should force the usage of `DisplayServerHeadless`, but results in `DisplayServerWindows` being used when in Windows. I haven't tried this on any other OS.

When stepping through `main.cpp` in a debugger this looks to be a regression caused by #88065, as `display_driver` always gets overwritten from `"headless"` to `"default"` here:

https://github.com/godotengine/godot/blob/e343dbbcc1030f04dc5833f1c19d267a17332ca9/main/main.cpp#L2634-L2638

### Steps to reproduce

1. Run the editor with `--headless --import` on any project.
2. Note how a window still appears.

### Minimal reproduction project (MRP)

N/A

**Comment by Kiisu-Master** [comment]
I just came here to report this as well. Had failing github actions when updating to rc1

**Comment by mihe** [comment]
> I noticed that when only using headless flag it works but when you add anything else like export or import it will try to use other display servers.

Anything that implies `--editor`, or is considered a `cmdline_tool` (like `--doctool`), will cause it to hit the offending code that I linked in the description, as seen here:

https://github.com/godotengine/godot/blob/e343dbbcc1030f04dc5833f1c19d267a17332ca9/main/main.cpp#L2563

**Comment by mihe** [comment]
I feel like the solution should be as simple as removing the `else { display_driver = "default"; }` part of the code I linked in the description, but I could be missing something.

**Comment by Kiisu-Master** [comment]
> Anything that implies `--editor`, or is considered a `cmdline_tool` (like `--doctool`), will cause it to hit the code that I linked in the description, as seen here:
> 
> https://github.com/godotengine/godot/blob/e343dbbcc1030f04dc5833f1c19d267a17332ca9/main/main.cpp#L2563

I haven't looked into it much but in the [github actions runs](https://github.com/MewPurPur/GodSVG/actions/runs/10100447599?pr=826) that I mentioned only one run failed and the other succeeded. This doesn't make a lot of sense to me, as both should be running on the same linux OS with the same command. Only difference is the export target and that one has WINE installed.

**Comment by HKunogi** [comment]
The error does happen on windows and on linux for me.

**Comment by jamie-pate** [comment]
Workaround for linux builds while you wait for rc2:
`xvfb-run --auto-servernum godot --rendering-driver opengl3`
(vulkan crashes for me with llvmpipe + xvfb)

**Comment by am4sten** [comment]
Running into the same issue and observing the segfault mentioned by @jamie-pate as well on 4.3-rc1.

However after some tweaking and playing with ubuntu versions in Github actions I managed to get some kind of stack trace that I hope could help. Interestingly enough I'm running my windows build and my linux build both under the same ubuntu VM and both with the --headless --verbose arguments. Only the X11 export fails as such on the same machine :

`Godot Engine v4.3.rc1.official.e343dbbcc - https://godotengine.org/
TextServer: Added interface "Dummy"
TextServer: Added interface "ICU / HarfBuzz / Graphite (Built-in)"
JoypadLinux: udev enabled and loaded successfully.
ERROR: X11 Display is not available
   at: DisplayServerX11 (platform/linuxbsd/x11/display_server_x11.cpp:5950)
Error: Can't open display: 
ERROR: Can't connect to a Wayland display.
   at: init (platform/linuxbsd/wayland/wayland_thread.cpp:3686)
ERROR: Could not initialize the Wayland thread.
   at: DisplayServerWayland (platform/linuxbsd/wayland/display_server_wayland.cpp:1337)
ERROR: Can't create the Wayland display server.
   at: create_func (platform/linuxbsd/wayland/display_server_wayland.cpp:1311)
ERROR: Unable to create DisplayServer, all display drivers failed.
Use "--headless" command line argument to run the engine in headless mode if this is desired (e.g. for continuous integration).
   at: setup2 (main/main.cpp:2760)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools11BucketLargeE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:[17](https://github.com/Memory-Shift-Studio/veteran-impact/actions/runs/10119557652/job/27988167629#step:7:18)0)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools12BucketMediumE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:170)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools11BucketSmallE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:170)
ERROR: BUG: Unreferenced static string to 0: TextServerDummy
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: IPUnix
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: ResourceFormatImporterSaver
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: MovieWriterMJPEG
   at: unref (core/string/string_name.cpp:1[29](https://github.com/Memory-Shift-Studio/veteran-impact/actions/runs/10119557652/job/27988167629#step:7:30))
ERROR: BUG: Unreferenced static string to 0: interface_added
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: TextServerAdvanced
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: ImageFormatLoader
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: ResourceFormatSaver
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: GDScriptEditorTranslationParserPlugin
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: MovieWriterPNGWAV
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: ResourceFormatLoader
   at: unref (core/string/string_name.cpp:129)
ERROR: BUG: Unreferenced static string to 0: ScriptLanguage
   at: unref (core/string/string_name.cpp:129)`
